sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/partitioning.scala

```diff
@@ -38,3 +38,11 @@ case class SortPartitions(sortExpressions: Seq[SortOrder], child: LogicalPlan)
  */
 case class RepartitionByExpression(partitionExpressions: Seq[Expression], child: LogicalPlan)
   extends RedistributeData
+
+/**
+ * This method repartitions data using [[Expression]]s into `numPartitions`. If numPartitions is
+ * less than zero, a default is used. Otherwise this behaves identically to RepartitionByExpression.
+ */
+case class PartitionByExpression(partitionExpressions: Seq[Expression],
+                                 child: LogicalPlan, numPartitions: Int = -1)
+  extends RedistributeData
```

Comment:
How about we just use `RepartitionByExpression`? Since users do not use classes in catalyst directly, it is fine to change the signature of the constructor. For the type of `numPartitions`, how about `Option[Int]`? If it is `None`, the number of partition is not specified. We can add a `require` clause to make sure this value is larger than 0 if `numPartitions` is a `Some`.

btw, regarding the format, the following one is preferred.
```
case class PartitionByExpression(
    partitionExpressions: Seq[Expression],
    numPartitions: Option[Int] = None,
    child: LogicalPlan)
  extends RedistributeData
```